### PR TITLE
framework/st_things : Add st_things fota update features

### PR DESCRIPTION
**Pull request by JeongJunSik**
1. Add CONFIG_ST_THINGS_FOTA configuration to update artik053 firmware
2. Boot-time generation of json file on smartfs
   previously, it was copied from ROMFS
3. st-things internal API uses only filename, not fullpath name

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
framework/src/st_things/things_stack/fota/fmwup_util_http.c:140: ERROR: [SPC_M_KWD] space required before the open brace '{'
framework/src/st_things/things_stack/fota/fmwup_util_http.c:148: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
framework/src/st_things/things_stack/fota/fmwup_util_http.c:190: ERROR: [BRC_M_FTN] open brace '{' following function declarations go on the next line
framework/src/st_things/things_stack/fota/fmwup_util_http.c:265: ERROR: [SPC_M_KWD] space required before the open brace '{'
framework/src/st_things/things_stack/fota/fmwup_util_http.c:309: ERROR: [SPC_M_KWD] space required before the open brace '{'

**Comment by seinfra** [comment]
Code Rule Check OK.

File: framework/src/st_things/Kconfig [hunk 1]
```diff
@@ -44,4 +44,9 @@ config ST_THINGS_STG_MODE
 	---help---
 		Build ST Things for STG Server
 
+config ST_THINGS_FOTA
+	bool "Enable ST Things FOTA Service"
+	default n
+	---help---
```
Comment (sunghan-chang) [inline framework/src/st_things/Kconfig]: You should add a dependancy with FOTA_HAL.

File: framework/src/st_things/things_stack/fota/Make.defs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+###########################################################################
+#
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/Make.defs]: 2018

File: framework/src/st_things/things_stack/fota/fmwup_api.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
+/* ****************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_api.c]: 2018

File: framework/src/st_things/things_stack/fota/fmwup_api.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+	fmwup_data_s *fmwup_data;
+	fmwup_data = fmwup_data_get_properties();
+
+	char *update;
+	update = fmwup_data_get_update_string(fmwup_data->update);
+
+	rep->things_set_int_value(rep, FIRMWARE_PROPERTY_UPDATE, update);
+	rep->things_set_value(rep, FIRMWARE_PROPERTY_UPDATE_TIME, fmwup_data->update_time);
+	rep->things_set_int_value(rep, FIRMWARE_PROPERTY_STATE, (int64_t) fmwup_data->state);
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_api.c]: remove a space after typecasing.

File: framework/src/st_things/things_stack/fota/fmwup_api.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	rep->things_set_int_value(rep, FIRMWARE_PROPERTY_RESULT, (int64_t) fmwup_data->result);
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_api.c]: remove a space after typecasing.

File: framework/src/st_things/things_stack/fota/fmwup_api.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
+/* ****************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_api.h]: 2018

File: framework/src/st_things/things_stack/fota/fmwup_api.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * @brief Enumeration for XXX.
+ */
+typedef enum {
+	FMWUP_ERROR_NONE = 0,			  /**< Successful */
+	FMWUP_ERROR_INVALID_PARAMETER = ST_THINGS_ERROR_INVALID_PARAMETER,
+	/**< Invalid parameter (If parameter is null or empty)*/
+	FMWUP_ERROR_OPERATION_FAILED = ST_THINGS_ERROR_OPERATION_FAILED,/**< Operation Failed */
+	FMWUP_ERROR_MEMORY_ERROR = -3,		  /**< Allocation Failed */
+} fmwup_things_error_e;
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_api.h]: We use "_t" postfix for typedef.

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 8]
```diff
@@ -0,0 +1,426 @@
+/* ****************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: Please check all of year what you add new file .

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+static pthread_mutex_t g_data_key_manager_mutex = PTHREAD_MUTEX_INITIALIZER;
+static const char *key_manager_original_json_str = "{\n\
+}";
+
+cJSON *root = NULL;
+
+int64_t fmwup_data_get_update_int64(const char *str)
+{
+	THINGS_LOG_D(THINGS_DEBUG, TAG, "str : %s", str);
+	if (strcmp(FIRMWARE_PROPERTY_UPDATE_INIT, str) == 0) {
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: strcmp will cause a warning on SVACE.

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+	} else if (strcmp(FIRMWARE_PROPERTY_UPDATE_UPDATE, str) == 0) {
+		return FMWUP_COMMAND_UPDATE;
+	} else if (strcmp(FIRMWARE_PROPERTY_UPDATE_DOWNLOADUPDATE, str) == 0) {
+		return FMWUP_COMMAND_DOWNLOADUPDATE;
+	} else {
+		return FMWUP_COMMAND_UNKNOWN;
+	}
+}
+
+char *fmwup_data_get_update_string(int64_t cmd)
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: include <stdint.h> to use int64_t

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+}";
+
+cJSON *root = NULL;
+
+int64_t fmwup_data_get_update_int64(const char *str)
+{
+	THINGS_LOG_D(THINGS_DEBUG, TAG, "str : %s", str);
+	if (strcmp(FIRMWARE_PROPERTY_UPDATE_INIT, str) == 0) {
+		return FMWUP_COMMAND_INIT;
+	} else if (strcmp(FIRMWARE_PROPERTY_UPDATE_CHECK, str) == 0) {
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: include <string.h> to use string api

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+		if (name) {
+			THINGS_LOG_D(THINGS_ERROR, TAG, "name [%s]", name);
+		}
+		if (data) {
+			THINGS_LOG_D(THINGS_ERROR, TAG, "data [%s]", data);
+		}
+		return -1;
+	}
+
+	if (root == NULL) {
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: include <sys/types.h> to use NULL

File: framework/src/st_things/things_stack/fota/fmwup_util_data.c [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+			   || (strcmp(name, FIRMWARE_PROPERTY_VENDER) == 0)
+			   || (strcmp(name, FIRMWARE_PROPERTY_MODEL) == 0)
+			  ) {
+		return platform_get_data(name);
+	} else {
+		THINGS_LOG_D(THINGS_DEBUG, TAG, "Invalid key[%s]", name);
+		return NULL;
+	}
+
+	return NULL;
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.c]: This is unreachable.

File: framework/src/st_things/things_stack/fota/fmwup_util_data.h [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#ifndef FMWUP_UTIL_DATA_H_
+#define FMWUP_UTIL_DATA_H_
+
+#include <stdbool.h>
+
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.h]: include <stdint.h>

File: framework/src/st_things/things_stack/fota/fmwup_util_data.h [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#ifndef FMWUP_UTIL_DATA_H_
+#define FMWUP_UTIL_DATA_H_
+
+#include <stdbool.h>
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.h]: There is no use for bool. I think you don't need to include this.

File: framework/src/st_things/things_stack/fota/fmwup_util_data.h [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+	char *package_uri;			//[RW][M] Firmware package URI where the package located
+
+	/* Device Information */
+	char *manufacturer;			//[R][O] Device vendor identifier
+	char *model_name;			//[R][O] Device model identifier
+	char *current_version;		//[R][M] Current firmware version
+
+} fmwup_data_s;
+
+int64_t fmwup_data_get_update_int64(const char *);
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_data.h]: coding rule violation.
[M11] In function prototypes, include parameter names with their data types and return type.

[Example-Do]
```
extern int function(struct type *ex);
```
[Example-Don't]
```
extern int function(struct type *);
```

File: framework/src/st_things/things_stack/fota/fmwup_util_http.c [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+	"KZsZ6YeH2Z81BauRj218kQKBgQCUV/DgKP2985xDTT79N08jUo3hTP5MVYCCuj/+\r\n"
+	"UeEw1TvZcx3LJby7P6Xad6a1/BqveaGyFKIfEFIaBUBItk801sDDpDaYc4gL00Xc\r\n"
+	"7lFuBHOZkxJYlss5QrGpuOEl9ZwUt5IrFLBdYaKqNHzNVC1pCPfb/JyH6Dr2HUxq\r\n"
+	"gxUwAQKBgQCcU6G2L8AG9d9c0UpOyL1tMvFe5Ttw0KjlQVdsh1MP6yigYo9DYuwu\r\n"
+	"bHFVW2r0dBTqegP2/KTOxKzaHfC1qf0RGDsUoJCNJrd1cwoCLG8P2EF4w3OBrKqv\r\n"
+	"8u4ytY0F+Vlanj5lm3TaoHSVF1+NWPyOTiwevIECGKwSxvlki4fDAA==\r\n"
+	"-----END RSA PRIVATE KEY-----\r\n";
+
+struct http_client_ssl_config_t g_config = {
+	(char *)c_ca_crt_rsa, (char *)c_cli_crt_rsa, (char *)c_cli_key_rsa,
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_http.c]: how about below for readability,
struct g_config = {
 A,
 B,
 C,
...
};

File: framework/src/st_things/things_stack/fota/fmwup_util_http.c [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+	sizeof(c_ca_crt_rsa), sizeof(c_cli_crt_rsa), sizeof(c_cli_key_rsa), 2
+};
+
+static int g_https;
+static bool g_finish = false;
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_http.c]: include <stdbool.h>

File: framework/src/st_things/things_stack/fota/fmwup_util_http.c [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+static void callback(struct http_client_response_t *response)
+{
+	if (response->status != 200) {
+		is_link_fail = true;
+		THINGS_LOG_D(THINGS_ERROR, TAG, "recv callback status %d", response->status);
+		return;
+	}
+
+	if (download_state == FOTA_DOWNLOAD_STATE_JSON) {
+		strncpy(json_str + recv_size, response->entity, response->entity_len);
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_http.c]: include <string.h>

File: framework/src/st_things/things_stack/fota/fmwup_util_http.c [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+	/* before sending request by sync function,
+	 * must initialize response structure
+	 */
+	if (http_client_send_request_async(&request, ssl_config, (wget_callback_t)callback)) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "fail to send request");
+		goto release_out;
+	}
+	/* sleep for end request */
+	while (request.async_flag > 0) {
+		usleep(100000);
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_http.c]: include <unistd.h>

File: framework/src/st_things/things_stack/fota/fmwup_util_http.h [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <tinyara/config.h>
+#include <protocols/webserver/http_keyvalue_list.h>
+
+#ifndef FMWUP_UTIL_HTTP_H_
```
Comment (sunghan-chang) [inline framework/src/st_things/things_stack/fota/fmwup_util_http.h]: 1. exchange locations of conditional and including header
```
#ifndef
#define

#include
```

2. Why do you include those headers in here?>

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Pull request merged**